Remove getitem special handling in the partitioner

IvanYashchuk opened the pull request:
This special handling of getitem unnecessary splits fusions at functions with tuple outputs.

Example script:
```py
import torch
from torch.fx.passes.infra.partitioner import CapabilityBasedPartitioner
from torch._prims.nvfuser_executor import NvfuserPrimOperatorSupport
from torch.fx.experimental.proxy_tensor import make_fx

def func(x):
    xx = torch.ops.nvprims.add(x, 1)
    var, mean = torch.ops.nvprims.var_mean(x, correction=0)
    var_cos = torch.ops.nvprims.cos(var)
    mean_sin = torch.ops.nvprims.sin(mean)
    return torch.ops.nvprims.add(var_cos, mean_sin)

a = torch.randn(5, 3, 3, device="cuda")
gm = make_fx(func)(a)
gm.graph.print_tabular()

supported_ops = NvfuserPrimOperatorSupport()
partitioner = CapabilityBasedPartitioner(
    gm, supported_ops, allows_single_node_partition=False
)
partitions = partitioner.propose_partitions()
print(partitions)
partitioned_graph = partitioner.fuse_partitions(partitions)
partitioned_graph.graph.print_tabular()
```
Output on master:
```py
opcode         name       target                       args              kwargs
-------------  ---------  ---------------------------  ----------------  -----------------
placeholder    x_1        x_1                          ()                {}
call_function  add        nvprims.add.default          (x_1, 1)          {}
call_function  var_mean   nvprims.var_mean.main        (x_1, [0, 1, 2])  {'correction': 0}
call_function  getitem    <built-in function getitem>  (var_mean, 0)     {}
call_function  getitem_1  <built-in function getitem>  (var_mean, 1)     {}
call_function  cos        nvprims.cos.default          (getitem,)        {}
call_function  sin        nvprims.sin.default          (getitem_1,)      {}
call_function  add_1      nvprims.add.default          (cos, sin)        {}
output         output     output                       (add_1,)          {}
[{cos, sin, add_1}, {var_mean, add, getitem, getitem_1}]
opcode         name       target                       args                    kwargs
-------------  ---------  ---------------------------  ----------------------  --------
placeholder    x_1        x_1                          ()                      {}
call_module    fused_1    fused_1                      (x_1,)                  {}
call_function  getitem_2  <built-in function getitem>  (fused_1, 0)            {}
call_function  getitem_3  <built-in function getitem>  (fused_1, 1)            {}
call_module    fused_0    fused_0                      (getitem_2, getitem_3)  {}
output         output     output                       (fused_0,)              {}
```
Output with this PR:
```
[{var_mean, add_1, cos, sin, add, getitem_1, getitem}]
opcode       name     target    args        kwargs
-----------  -------  --------  ----------  --------
placeholder  x_1      x_1       ()          {}
call_module  fused_0  fused_0   (x_1,)      {}
output       output   output    (fused_0,)  {}
```

IvanYashchuk committed:
Remove getitem special handling in the partitioner

This special handling of getitem unnecessary splits fusions at functions
with tuple outputs.

jjsjann123 commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -148,26 +142,6 @@ def merge_single_node(node: Node, id: int):
                     # this is a no-op
                     maybe_merge_partition(self_id, other_id)
 
-        # post processing to re-assign "getitem" nodes into upstream partition
```
nitpick, should we remove this part?

This was supposed to fuse getitem to producer, which is a fusion, before we have `getitem` disabled in the line you commented above. With that logic removed, I believe this section would remove getitem node at the beginning of the fusion, if it comes from an unfused node.

IvanYashchuk replied:
Keeping this code works with the example I added to the PR description, but I was hitting the assert in `merge_single_node` in one benchmark model. I should extract that failing graph portion for testing.

jjsjann123 replied:
Got ya. merge_single_node might be asserting too aggressively as well.

https://github.com/pytorch/pytorch/blob/8393213a04542b1be58170a79afe5312f1871c72/torch/fx/passes/infra/partitioner.py#L108

This assert should be removed along with you update on `getitem` logic.

IvanYashchuk replied:
Keeping this code `python -m pytest test/test_fx_passes.py -k "test_partitioner_fn_" -vvv` fails with the assert
```py
Traceback (most recent call last):
  File "/home/iyashchuk/dev/pytorch/master/test/test_fx_passes.py", line 224, in test_partitioner
    partitions = partitioner.propose_partitions()
  File "/home/iyashchuk/dev/pytorch/master/torch/fx/passes/infra/partitioner.py", line 162, in propose_partitions
    merge_single_node(node, id)
  File "/home/iyashchuk/dev/pytorch/master/torch/fx/passes/infra/partitioner.py", line 108, in merge_single_node
    assert node not in assignment
AssertionError
```
Removing this code the test still fails (as in CI) but the failure is expected now:
```py
____ TestFXGraphPasses.test_partitioner_fn_<function TestPartitionFunctions_forward13 at 0x7f08e8fdfd00>_expected_partition_[['add_2', 'add_1', 'add']] ____
Traceback (most recent call last):
  File "/home/iyashchuk/dev/pytorch/master/test/test_fx_passes.py", line 229, in test_partitioner
    assert set(partitions_name[i]) == set(expected_partition[i])
AssertionError: assert {'add', 'getitem_2', 'getitem', 'getitem_1', 'getitem_3', 'add_2', 'add_1'} == {'add_2', 'add_1', 'add'}
  Extra items in the left set:
  'getitem_2'
  'getitem_3'
  'getitem_1'
  'getitem'
  Full diff:
  - {'add_2', 'add_1', 'add'}
  + {'add', 'getitem_2', 'getitem', 'getitem_1', 'getitem_3', 'add_2', 'add_1'}
```
The failing test was added in `https://github.com/pytorch/pytorch/pull/86713` 😉

jjsjann123 replied:
I was suggesting keeping this code and remove the assert.

jjsjann123 replied:
Oops, that's surprising to me. :laughing:

How can we have `getitem` spanning acorss multiple partition? I think there's just a small bug somewhere. Looks like it's attempting to clear `getitem` from the original partition... But somehow that wasn't done right. I can take a quick look at this afterwards.

jjsjann123 reviewed (approved):
LGTM.

IvanYashchuk committed:
Fix getitem test

IvanYashchuk committed:
getitem is supported only if node to be unpacked is supported by nvFuser

jjsjann123 commented on torch/_prims/nvfuser_executor.py:
```diff
@@ -237,6 +237,10 @@ def is_node_supported(self, submodules, node: torch.fx.Node) -> bool:
+        if "getitem" in node.name:
+            # Check if the node unpacks a tuple from a supported node
+            node_to_unpack = node.args[0]
+            return self.is_node_supported(submodules, node_to_unpack)
```
I'm slightly leaning towards clean up `getitem` at the beginning of the graph in partitioner, instead of having more complicated logic in op support query. For now it works fine, since partitioner goes from consumer to producer and we don't really fuse anything as we propose partition group. But that implementation could change :frowning:

IvanYashchuk replied:
It's of course a matter of taste. I think special casing `getitem` and processing it differently may cause bugs. There are two situations with getitem it's either unpack a node that's supported or not, and we can't unconditionally either reject or accept getitem. Previously we have already relied on the partitioner (always rejecting getitem) to do the right thing and it didn't.

SherlockNoMad commented on test/test_fx_passes.py:
```diff
@@ -214,7 +214,7 @@ class TestFXGraphPasses(JitTestCase):
 
         # 5 getitem special case
-        (TestPartitionFunctions.forward13, [["add_2", "add_1", "add"]]),
+        (TestPartitionFunctions.forward13, [["add_2", "add_1", "add", "getitem", "getitem_1", "getitem_2", "getitem_3"]]),
```
Hi @IvanYashchuk, I think this is an unexpected change. 
getitem node should always be partitioned together with its producer node. 
Across the stack, we have an implicit assumption that module's input and output must be tensor type. This is why we have the special handling logic in the first place.

jjsjann123 replied:
This doesn't look right even for this PR. `getitem` is produced by `split`, which isn't supported by the fusion node?

SherlockNoMad commented:
Given the example, getitem_2, getitem_3 should be partitioned into fused_1.

SherlockNoMad reviewed (changes requested):
getitem should always be partitioned together with its producer node.

jjsjann123 commented:
> getitem should always be partitioned together with its producer node.

@SherlockNoMad The problem here is that the original logic special case `getitem` after fusion partition has been proposed. Which resulted in us always segment the graph across `getitem` nodes.
i.e. if you have a `var_mean` where it's output is used by another fusion-supported op, the `getitem` node that's used to unpack the tuple output from `var_mean` stops us from fusing the two nodes.

Changes in this PR is supposed to fix that.

IvanYashchuk committed:
mypy fix

IvanYashchuk committed:
Accept getitem node only if its producer is supported

IvanYashchuk commented:
> getitem should always be partitioned together with its producer node.

I agree with you, but the previous code was doing that too aggressively and incorrectly. `getitem` shouldn't be unconditionally accepted nor rejected, it should be tied to its producer when deciding whether it's supported or not.

Please take a look at the latest changes.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into getitem-partitioner

jjsjann123 commented:
> > getitem should always be partitioned together with its producer node.
> 
> I agree with you, but the previous code was doing that too aggressively and incorrectly. `getitem` shouldn't be unconditionally accepted nor rejected, it should be tied to its producer when deciding whether it's supported or not.
> 
> Please take a look at the latest changes.

On this topic, could you consider changes like this one?
 https://github.com/csarofeen/pytorch/pull/2099/commits/9e9934e3e10ef098de68c4ea803856630f21eb5e
 
 We are merging things as is provided by op support list, and at the end of fusion, we clean up `getitem` node and keep them with their producer.
 
 Sorry that my previous refactor leaves the code in an ugly state. :stuck_out_tongue_closed_eyes:

jjsjann123 commented:
> On this topic, could you consider changes like this one? [csarofeen@9e9934e](https://github.com/csarofeen/pytorch/commit/9e9934e3e10ef098de68c4ea803856630f21eb5e)

Seems like we are not the sole user of fx partitioner. https://github.com/pytorch/pytorch/pull/87007

if `getitem` is relied by multiple backends, maybe it's a better that we keep the logic on merging `getitem` in op support simpler and have a common post processing pass to clean it up. I'm shamelessly promoting the patch I have in the commit above again :)

I'm not sure if the discussion here concerns you at all @wschin, would an `getitem` node at the beginning of partition break your parser?

wschin commented:
> > On this topic, could you consider changes like this one? [csarofeen@9e9934e](https://github.com/csarofeen/pytorch/commit/9e9934e3e10ef098de68c4ea803856630f21eb5e)
> 
> Seems like we are not the sole user of fx partitioner. #87007
> 
> if `getitem` is relied by multiple backends, maybe it's a better that we keep the logic on merging `getitem` in op support simpler and have a common post processing pass to clean it up. I'm shamelessly promoting the patch I have in the commit above again :)
> 
> I'm not sure if the discussion here concerns you at all @wschin, would an `getitem` node at the beginning of partition break your parser?

Some `getitem`s in the beginning and at the end  should be fine. I just hope most of the computation can be partitioned into a single torch.fx.GraphModule. Having non-computation ops before and after the major computation should be fine; of course, those non-computation ops should be as less as possible. Many thanks!

IvanYashchuk committed:
Bring back nodes reassignment code

IvanYashchuk committed:
Add test with var_mean

IvanYashchuk committed:
Fix getitem handling

Placeholder nodes called "getitem_XXX" were incorrectly dropped from the
graph.

jjsjann123 committed:
updated getitem handling & new tests

jjsjann123 committed:
fixing import

jjsjann123 committed:
fixing tests

IvanYashchuk committed:
Merge pull request #4 from jjsjann123/getitem_patch

Getitem patch

IvanYashchuk commented:
@SherlockNoMad could you please take another look at the proposed changes?

IvanYashchuk committed:
Update nvfuser_executor.py

SherlockNoMad commented on torch/fx/passes/infra/partitioner.py:
```diff
@@ -110,13 +104,17 @@ def dfs_find_cycle(node):
 
-            assignment[node] = id
-            if id not in partitions_by_id:
+            if id is None:
```
Do we really have case where id is None？ 
Looking at the call site for this function, it doesn't seem to have any None case... 

If this cannot be None, let assert id is not None.

jjsjann123 replied:
Yes, this happens naturally when `getitem` is marked as `supported` by backends.

Later during the special handling of `getitem`, where we merge each `getitem` calls to its producer, we could run into cases where the producer is not supported by backends, but we accidentally merged `getitem` into the fusion. We'll ended up pulling these nodes out and that's when `id` here would be `None`.

https://github.com/pytorch/pytorch/blob/7d3fbd7603adcb0421373f5b1924b5adfd8f574a/torch/fx/passes/infra/partitioner.py#L167

This is checked in test case `forward13`

SherlockNoMad reviewed (approved):
LGTM, expect for the minor comment.

jjsjann123 commented:
Thanks to @SherlockNoMad 's stamp.

Failure seems to be on CI nodes and unrelated. Let's rebase to a stable commit and merge it! :tada: :tada: :tada:

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into getitem-partitioner

IvanYashchuk commented:
@pytorchbot merge -g
